modules/storage-api/src/main/java/org/apache/ignite/internal/storage/util/StorageOperation.java

```diff
@@ -25,16 +25,55 @@
  * Storage operations.
  */
 interface StorageOperation {
+    /**
+     * Returns future completion of the operation.
+     */
+    CompletableFuture<Void> operationFuture();
+
+    /**
+     * Return {@code true} if the operation is the final.
+     */
+    boolean isFinalOperation();
+
+    /**
+     * Marks the operation as final.
+     */
+    void markFinalOperation();
+
+    /**
+     * Abstract operation of the storage.
+     */
+    abstract class AbstractStorageOperation implements StorageOperation {
```

Comment:
If all implementations must extend abstract class now, why do you need an interface? Just replace it with abstract class.

Reply:
Why not, fixed it.